### ISSUE TITLE
[Question] After train, how to test our own environment?

### ISSUE THREAD

**Issue by zhou-ting-hub** [PKU-Alignment/omnisafe#306]
### Required prerequisites

- [X] I have read the documentation <https://omnisafe.readthedocs.io>.
- [X] I have searched the [Issue Tracker](https://github.com/PKU-Alignment/omnisafe/issues) and [Discussions](https://github.com/PKU-Alignment/omnisafe/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/PKU-Alignment/omnisafe/discussions/new).

### Questions

Thank you for your work. After I successfully run the following train code 

```
cd examples
python train_policy.py --algo CPO --env Custom0-v0
```

how to test next? 
 
Method1: 

![image](https://github.com/PKU-Alignment/omnisafe/assets/80151423/e573ebbb-6cfd-412a-a7be-ff6037709ff6)
we modified to `omnisafe eval ./examples/runs/CPO-{Custom0-v0}`

Method2:   run the file ./examples/evaluate_saved_policy.py

` LOG_DIR = /examples/runs/CPO-{Custom0-v0}/seed-000-2024-02-29-23-33-21`

**So  which is right or how do is right to test the trained model, what is the difference between Method1 and Method2?  Is the trained model saved in examples/runs/CPO-{Custom0-v0}/seed-000-2024-02-29-23-33-21? Thank you~**

![image](https://github.com/PKU-Alignment/omnisafe/assets/80151423/82c23fdb-89cb-4dd9-95f7-95fdb96994fb)

**Comment by Gaiejj** [comment]
Yes, the location where the experimental results are saved is ``examples/runs/CPO-{Custom0-v0}/seed-000-2024-02-29-23-33-21``.
In fact, both methods of evaluating trained policies are fine. The ``omnisafe eval`` command provides more extensive command line information, making it more suitable for beginners to use simply; while ``examples/evaluate_saved_policy.py`` offers an interface at the code level for evaluation, facilitating customization for users. For example, using loop logic (e.g. for xxx) to iterate and evaluate all results in a specific folder. 
If you encounter difficulties in the process of using these two methods for evaluation, feel free to continue providing feedback.

**Comment by zhou-ting-hub** [comment]
Thank you, to test, I also have tried  the file `./examples/train_from_custom_dict.py` and `train_from_yaml.py`, that has been solved, but I have another two problems:
Q1: When I use Method2: run the file `./examples/evaluate_saved_policy.py`

 LOG_DIR = /examples/runs/CPO-{Custom0-v0}/seed-000-2024-02-29-23-33-21

(1) For the same LOG_DIR, why are the results different? In theory, to evaluate the same model, the results shoule be the same.
(2) Does the evaluation use the current environment or just use the saved model in  LOG_DIR? 

Q2: To run the file `./examples/train_policy.py`, we want to use gpu, also the environment are equipped with the torchgpu, but the erro is as the following:


![1709804855092](https://github.com/PKU-Alignment/omnisafe/assets/80151423/9e254363-102f-4b5f-a344-036c47902f18)

**Comment by Gaiejj** [comment]
**Q1**
Yes, when we evaluate the trained policy, we only import the trained policy and use a randomly initialized environment. This will cause the results of each evaluation to be inconsistent. If you need the results of each evaluation to be consistent, you need to make the following change in `omnisafe/evaluator.py`:
```python
SEED=5 # for example
from omnisafe.utils.tools import seed_all
seed_all(seed=SEED)
```
Then in the method `__load_model_and_env`, after making the env by `self._env = make(**env_kwargs)`, add:
```python
self._env.set_seed(seed=SEED)
```
Please note, that to ensure the rigor of the evaluation, use a different random seed for the evaluation than the one used during training.

---
**Q2**
Please additionally specify the GPU id. e.g. `cuda:0`

---
**Q3**
- The original cost_limit is 25.00 in the algorithms `.yaml` file. You can try to set it to zero to obtain an expected result.
- Sure, the environment will be stopped and reset after truncation.

**Comment by zhou-ting-hub** [comment]
THANK YOU VERY MUCH! 
But we face a new problem,  we run train_policy.py or train_from_custom_dict.py:
Why the train results is the same after we modified our environment under the premise of the same random seed settings in CPO.yaml? 
For example, we modified the reward function in environment, the train results is the same; or modified  some variables range in environment, the train results is just a little difference in range, the train results is the same trend. 
In short, the training results did not further learn the modified environment but remained the original decision-making framework.

**Comment by Gaiejj** [comment]
I believe this is due to the issue with **the environment's random seed mechanism**. The environment officially supported by OmniSafe is Safety-Gymnasium, which is based on [Gymnasium](https://github.com/Farama-Foundation/Gymnasium), commonly used in the reinforcement learning community. In the random seed setting mechanism of Gymnasium, the environment **generates a series of random numbers based on the initial random seed**, which are used as seeds for subsequent resets, instead of using the same seed for every seed. For more details please refer to here: https://github.com/Farama-Foundation/Gymnasium/blob/main/gymnasium/utils/seeding.py

**Comment by zhou-ting-hub** [comment]
Thank you, but In our environment, we reference the seed seeting in simple-env.py as following:
```
    def reset(
        self,
        seed: int | None = None,
        options: dict[str, Any] | None = None,
    ) -> tuple[torch.Tensor, dict]:
        if seed is not None:
            self.set_seed(seed)
        obs = torch.as_tensor(self._observation_space.sample())
        self._count = 0
        return obs, {}

    def set_seed(self, seed: int) -> None:
        random.seed(seed)
```
Also there is seed=0 in CPO.yaml,  how it work?
If seed is not None: not run because the seed is None in def reset？
random.seed(seed) is the same meaning? when we change it into random.randint(1,10), no work. 
so what should we modify the seed setting in our environment or in other place to achieve different random reaults.

![image](https://github.com/PKU-Alignment/omnisafe/assets/80151423/fd2505a5-1ef0-4d28-999e-c770ab77d6a0)

**Comment by Gaiejj** [comment]
You can set a simple seed logic, such as adding 10 each time. That is, when you first reset the environment, the seed you pass in is 0, and for subsequent resets, you only need to pass in None, allowing the environment to automatically reset with seeds 10, 20, 30, and so on. Similarly, when your initial seed is 5, the environment will automatically reset in the order of 15, 25, 35, and so on. This logic can be easily implemented in the reset function.

**Comment by zhou-ting-hub** [comment]
There are two places related to seed, so which seed plays the effect?

1. In our environment, def reset() and def set _seed() is modified as following, 
 ```
   def reset(
        self,
        seed: int | None = None,
        options: dict[str, Any] | None = None,
    ) -> tuple[torch.Tensor, dict]:
        if seed is  None:
            self.set_seed(seed)
        obs = torch.as_tensor(self._observation_space.sample())
        self._count = 0
        return obs, {}

    def set_seed(self, seed: int) -> None:
        seed=random.randint(1,10)
        random.seed(seed)
```
According your advice, we modified the seed=random.randint(1,10),  as a result, we found that although seed is different in each episode such as seed=10,10,10,10,10 responding to five episodes, and seed is the same seed=10,10,10,10,10 when we  run next time,  so the train results is the same. 

There are some description in [https://blog.csdn.net/qq_42951560/article/details/112184965], so we delete  random.seed(seed). as a result, we found that although seed is different in each episode such as seed=2,5,3,6,7 responding to five episodes, but seed is the same seed=2,5,3,6,7 when we  run next time,  so the train results is still the same. 

The following is related to the  def set_seed(), but we do not find solve methods.
(1) In omnisafe\algorithms\base_algo.py,   
`self._init_env()`
(2)In omnisafe\algorithms\on_policy\base\policy_gradient.py, 
```
self._env: OnPolicyAdapter = OnPolicyAdapter(
            self._env_id,
            self._cfgs.train_cfgs.vector_env_nums,
            self._seed,
            self._cfgs,
        )
```
(3)In omnisafe\adapter\onpolicy_adapter.py,
`super().__init__(env_id, num_envs, seed, cfgs)`
(4)In omnisafe\adapter\online_adapter.py,
`self._env.set_seed(seed)`

3.  In omnisafe\algorithms\base_algo.py,  it used cfgs.seed that is seed:0 in CPO.yaml
```
        assert hasattr(cfgs, 'seed'), 'Please specify the seed in the config file.'
        self._seed: int = int(cfgs.seed) + distributed.get_rank() * 1000
        seed_all(self._seed)
```
we modified  the above as:
```
        assert hasattr(cfgs, 'seed'), 'Please specify the seed in the config file.'
        self._seed=random.randint(1,10)
        seed_all(self._seed)
```
we find that the train results can be differnt. We do not know if it is right?

**Comment by Gaiejj** [comment]
I think I need to clarify the meaning of the seed mechanism:
- For multiple episodes of environment interactions in a single training session, each episode's environment random seed should differ. As previously mentioned, a simple logic can be used to reset the environment with a different random seed each time.
- For multiple training sessions, the same seed should yield the same results. This is for the reproducibility of experiments.
As for your question, CPO.yaml's seed actually is playing an effect. You can modify it to run in a different initial random seed.

**Comment by zhou-ting-hub** [comment]
Thank you, I unsderstand your meaning.

- For multiple episodes of environment interactions in a single training session,  each episode's environment random seed should differ.  I think the seed is None can satisfy that. We do not need ro to reset the random seed with the above simple logic extraly.

- For multiple training sessions,  the same seed should yield the same results. **I think this is for the same environment, but when we modify the environment, the results is still the same, this is my key problem,** so I delete the seed_all(self._seed) In omnisafe\algorithms\base_algo.py, or modify the  CPO.yaml's seed  as you suggested, to achieve different results when next train.

**Comment by zhou-ting-hub** [comment]
The training Reward should increase, cost should decrease.

Q1:  Why is reward in a downward trend? Our goal is to minimize economic costs, so we set a negative value, such as` reward=-(self.price_e*self.P_buy+self.price_q*self.Q_buy)*1e-4` 

Q2: Does CPO only support one constraint? Our setting is   `cost=torch.as_tensor(max(max(0,self.Q_buy-Max_Q_buy),0-self.Q_buy)+max(max(0,self.P_buy-Max_P_buy),0-self.P_buy) `,  there are two constraints.  Is this related to the decrease in rewards?

Thank you for your reply!

**Comment by Gaiejj** [comment]
I'm sorry, but I'm not an expert in applying SafeRL to trading transactions. You need to focus on whether maximizing reward and minimizing cost can coexist simultaneously. For instance, in the Safety-Gymnasium supported by OmniSafe, specifically in SafetyPointGoal1-v0, maximizing reward (reaching the goal) and minimizing cost (avoiding collisions) can coexist, meaning the agent can choose a safe path to the goal. If the environment is designed to meet this condition, then it might be because the default parameters of CPO are not well suited to your task, and you can use `examples/benchmarks/run_experiment_grid.py` to search for the optimal hyperparameters.

OmniSafe's CPO currently does not support multiple constraints. You can try to handle this by summing up the two cost functions or taking their average, depending on their actual meanings.

**Comment by zhou-ting-hub** [comment]
Some problems about reward and cost learning curve:

we run the `train_from_custom_dict.py`, the epoch is set 4000, the results of `agent.plot()` and `agent.render() `are saved in `D:\3omnisafe-main\examples\runs\CPO-{Custom0-v0}\seed-001-2024-04-26-17-29-33`.

<img width="418" alt="57dccecd0223dffad08c70a9a810647" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/e5625ed4-fc07-467a-8ba4-6e6a8a0c6785">





Q1: In `def plot()` of `C:\Users\zyt\.conda\envs\omnisafegpu\Lib\site-packages\omnisafe\algo_wrapper.py`, the plot results zyt.png is saved in `D:\3omnisafe-main\examples\runs\CPO-{Custom0-v0}\seed-001-2024-04-26-17-29-33\zyt.png`

<img width="636" alt="1714139242170" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/7fc4c84f-5008-4a48-96be-dbcb9ed93a96">

<img width="480" alt="1714140059708" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/3c2f84a9-f08b-414b-b17b-a07cba9cb98c">

<img width="863" alt="1714141566248" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/ff825435-00f7-44c1-bccb-617ffb198240">


but when we run ` tensorboard --logdir   D:\3omnisafe-main\examples\runs\CPO-{Custom0-v0}\seed-001-2024-04-26-17-29-33\tb` , the tensorboard  results are as following:


![image](https://github.com/PKU-Alignment/omnisafe/assets/80151423/79c3d416-abef-40d1-865c-ab447a0fc6ab)

<img width="1244" alt="1714141697715" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/fd42613a-1df4-4fdb-a0de-80c1790f6b61">

we find that convergence curves in tensorboard  is the same with the `progess.csv` that reflects the training process as following:

![image](https://github.com/PKU-Alignment/omnisafe/assets/80151423/1ea50776-d411-41f2-bf90-46e72fbc8fb6)

![1714142064373](https://github.com/PKU-Alignment/omnisafe/assets/80151423/384644bf-f33e-4f4d-9b4f-c0ec9dc0d3cb)



**In conclusion, we find that convergence curves in tensorboard  is the same with the `progess.csv` that reflects the training process, but is different from the above zyt.png,  so what reflects of zyt.png obtained by  `agent.plot()` in `omnisafe\algo_wrapper.py`, it is not  the  convergence curve?**


Q2: In `def render()` of `C:\Users\zyt\.conda\envs\omnisafegpu\Lib\site-packages\omnisafe\evaluator.py`, the render results are saved in `D:\3omnisafe-main\examples\runs\CPO-{Custom0-v0}\seed-001-2024-04-26-17-29-33\video`

<img width="593" alt="1714140580696" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/dc407ba7-9d0b-49ca-8b29-c19dbca731d6">


<img width="599" alt="1714140263078" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/0e19e68f-0f41-439a-984b-bef17e64f7fe">

**we find the render results in video\epoch4000\result is terrible, why the training curve has been converged, the reward and cost values of render results are terrible than the  training curve? 
such as the cost of render results is still big as 17870, but  the cost of training curve in tensorboard has been converged to 60?  the reward of render results is still small as -648, but  the reward of training curve in tensorboard has been converged to -430?**


<img width="548" alt="1714140652595" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/e45d1669-72f4-4399-a069-125d2aaa7273">

**Comment by zhou-ting-hub** [comment]
Looking forward to your reply, thank you~

**Comment by zhou-ting-hub** [comment]
Looking forward to your reply about the above problems, thank you~

**Comment by Gaiejj** [comment]
I apologize for the late reply. I will address your questions one by one:
(1) **The curve data in `zyt.png` is consistent with that in TensorBoard**. As for the reason for the discrepancy in visual presentation, it is because TensorBoard automatically ignores excessively high outlier values. For instance, in the TensorBoard chart you showed, the EpCost scale is around 60-180; whereas in `zyt.png`, the `EpCost` scale is 0-14000. You only need to use the following code to set the display range for the axis in line 160 of `omnisafe/utils/plotter.py`.
```python
sub_figures[1].set_ylim(COST_LOWER, COST_UPPER)
sub_figures[0].set_ylim(REWARD_LOWER,REWARD_UPPER)
```

(2)
I noticed that your main concern is why the agent's performance during render() is inconsistent with training. Addressing your three questions, here are my explanations:

a. The original design intention of `render()` is to visualize evaluation results, so it serves both as evaluation and visualization. Evaluation includes two aspects: 1. The agent generates actions using a deterministic strategy, not a random strategy. 2. The random seed in the agent's evaluation environment is different from that in the training environment.

b.  If your evaluation results are very inconsistent with training, you might consider changing the deterministic strategy to a stochastic strategy, like:
```python
act = self._actor.predict(
    obs.reshape(
        -1,
        obs.shape[-1],  # to make sure the shape is (1, obs_dim)
    ),
    deterministic=False,
).reshape(
    -1,  # to make sure the shape is (act_dim,)
)
```

or carefully check whether the environment imported during `render()` is consistent with the training environment.
2 / 2

**Comment by zhou-ting-hub** [comment]
Thank you, for (2)b, we modified that` in def render(） in evaluator.py` as follows, but run  `evaluator.render(num_episodes=1) `results in `evaluate_saved_policy.py` is still very inconsistent with training.  and modified that` in def evaluate(） in evaluator.py`,    but run `evaluator.evaluate(num_episodes=1)` results is also terrible.


<img width="404" alt="1715661128284" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/499c3b8a-7d5a-4937-97ee-89199d80c49c">

<img width="358" alt="1715661230195" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/e3fd1002-9c55-44ea-8b6a-30d2274a74f6">



and for (2)a, you said '' The random seed in the agent's evaluation environment is different from that in the training environment. "
we set the seed in  `evaluator.py` the seed=0 is the same as seed =0 in `CPO.YAML` that used in training. but run  `evaluator.render(num_episodes=1)`  and `evaluator.evaluate(num_episodes=1) ` in `evaluate_saved_policy.py` results are both terrible.

<img width="346" alt="1715660617330(1)" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/89169b47-5881-443d-99c6-0759ef968196">
<img width="357" alt="1715660723029" src="https://github.com/PKU-Alignment/omnisafe/assets/80151423/8b20af97-8b67-470d-af55-ffa1fb9b5a29">
